ksql-engine/src/main/java/io/confluent/ksql/function/InternalFunctionRegistry.java

```diff
@@ -34,31 +34,33 @@
 import io.confluent.ksql.function.udf.string.LenKudf;
 import io.confluent.ksql.function.udf.string.TrimKudf;
 import io.confluent.ksql.function.udf.string.UCaseKudf;
+import io.confluent.ksql.function.udtf.array.ExplodeFunctionFactory;
 import io.confluent.ksql.name.FunctionName;
+import io.confluent.ksql.util.KsqlConstants;
 import io.confluent.ksql.util.KsqlException;
 import java.util.ArrayList;
 import java.util.Collections;
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
-import java.util.concurrent.ConcurrentHashMap;
 import javax.annotation.concurrent.ThreadSafe;
 import org.apache.kafka.connect.data.Schema;
 import org.apache.kafka.connect.data.SchemaBuilder;
 
 @ThreadSafe
 public class InternalFunctionRegistry implements MutableFunctionRegistry {
 
-  private final Object lock = new Object();
-  private final Map<String, UdfFactory> udfs = new ConcurrentHashMap<>();
-  private final Map<String, AggregateFunctionFactory> udafs = new ConcurrentHashMap<>();
+  private final Map<String, UdfFactory> udfs = new HashMap<>();
+  private final Map<String, AggregateFunctionFactory> udafs = new HashMap<>();
+  private final Map<String, TableFunctionFactory> udtfs = new HashMap<>();
   private final FunctionNameValidator functionNameValidator = new FunctionNameValidator();
 
   public InternalFunctionRegistry() {
     new BuiltInInitializer(this).init();
   }
 
-  public UdfFactory getUdfFactory(final String functionName) {
+  public synchronized UdfFactory getUdfFactory(final String functionName) {
```

Comment:
IMHO switching the locking model on this class is an unrelated change, though I know @agavra commented on it, and would be better done as a separate PR, where we can have a more targeted discussion.

Reply:
https://github.com/confluentinc/ksql/issues/3674